### ISSUE TITLE
Bug: Algorithm problem in SimpleSOMMapper

### ISSUE THREAD

**Issue by Soletmons** [PyMVPA/PyMVPA#342]
I have been experimenting with the SimpleSOMMapper class, and came across what I think is a severe algorithm problem (please correct me if I'm wrong on this).

Currently, when training the SOM, the unit_delta to apply to each weight vector in the Kohonen layer is cumulatively calculated for each input training vector. This is then applied to the Kohonen layer only after every training vector has been considered. (see code snippet below)

This methodology has the unfortunate problem of creating very large numbers in the Kohonen layer. If a vector of weights is say 1 unit above all of a set of 1000 training vectors, the effect of one iteration will be to reduce the weights by 1000, where the desired effect is to reduce them by 1 (subject to learning rates/influence kernels). The next iteration, the weights will be increased by 1000*1000, and so on. Over many iterations, the Kohonen layer is driven to unreasonable values, breaking the algorithm. I believe this hasn't been spotted because the effect generally only becomes apparent for medium+ size sets of training vectors or iteration numbers, and the colours example used in the documentation is small.

The solution to this is to update the Kohonen layer each time a single training vector is considered (I'm happy to alter the code for this if you'd like).

It is also worth pointing out that the current implementation is different from some other versions of SOMs, in that often only one training vector is selected at random to apply on each iteration (here every training vector is used). Not necessarily a problem but important to say in the documentation.

Relevant code from SimpleSOMMapper:

        # for all iterations
        for it in xrange(1, self.niter + 1):
            # compute the neighborhood impact kernel for this iteration
            # has to be recomputed since kernel shrinks over time
            k = self._compute_influence_kernel(it, dqd)

            # for all training vectors
            for s in samples:
                # determine closest unit (as element coordinate)
                b = self._get_bmu(s)
                # train all units at once by unfolding the kernel (from the
                # single quadrant that is precomputed), cutting it to the
                # right shape and simply multiply it to the difference of target
                # and all unit weights....
                infl = np.vstack((
                        np.hstack((
                            # upper left
                            k[b[0]:0:-1, b[1]:0:-1],
                            # upper right
                            k[b[0]:0:-1, :self.kshape[1] - b[1]])),
                        np.hstack((
                            # lower left
                            k[:self.kshape[0] - b[0], b[1]:0:-1],
                            # lower right
                            k[:self.kshape[0] - b[0], :self.kshape[1] - b[1]]))
                               ))
                unit_deltas += infl[:, :, np.newaxis] * (s - self._K)

            # apply cumulative unit deltas
            self._K += unit_deltas

**Comment by nno** [comment]
Thanks for the report, and for the offer to provide code that would address this issue. I'm tagging @hanke because git-blame lists him as the main author of this function. 

To address this, it would be great if you could provide a pull request that:

- implements the required changes in ``mappers/som.py``.
- adds a unit test to ``tests/test_som.py`` that fails using the current code, and passes using your new code.

(If you are not experienced using git, you could also send a patch to the developers by email directly).

**Comment by hanke** [comment]
Hi,

thanks for bringing this to my attention. I agree that the current
implementation is primitive and your proposed changes sound good. I'd be
happy to review a PR that merges them into PyMVPA.

Cheers and thanks,

Michael


On Wed, Aug 12, 2015 at 11:34 AM, Nikolaas N. Oosterhof <
notifications@github.com> wrote:

> Thanks for the report, and for the offer to provide code that would
> address this issue. I'm tagging @hanke <https://github.com/hanke> because
> git-blame lists him as the main author of this function.
>
> To address this, it would be great if you could provide a pull request
> that:
>
>    - implements the required changes in mappers/som.py.
>    - adds a unit test to tests/test_som.py that fails using the current
>    code, and passes using your new code.
>
> (If you are not experienced using git, you could also send a patch to the
> developers by email directly).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PyMVPA/PyMVPA/issues/342#issuecomment-130236619>.
>



-- 
Michael Hanke
http://mih.voxindeserto.de

**Comment by Soletmons** [comment]
Hi both,

Thanks for getting back. I'd be glad to help if I can (though I'm a complete beginner to github).

I will try to work out how to set up a pull request with the changes. I have also been using this algorithm for a project I've been working on, and have implemented periodic boundary conditions on the Kohonen layer. Would including this interest you? It helps stop different training vectors clustering into the corners of the layer.

In summary, I can:
- Fix the bug in the first comment
- Include periodic boundaries to the Kohonen Layer (if you would like)

Cheers,
David

**Comment by nno** [comment]
> In summary, I can:
> 
> Fix the bug in the first comment
> Include periodic boundaries to the Kohonen Layer (if you would like)

That would be great. Please also add some test code (in ``tests/test_som.py``), preferably in such a manner that the original code fails the test and your new code passes. 

In that way the test suite can help us in being more confident that the code fixes the issue, and detect it when new changes would break anything.

**Comment by Soletmons** [comment]
I've just added the pull request. Hope I made sense of github well enough.

**Comment by nno** [comment]
Addressed by #355.
